framework/src/variables/MooseVariableFE.C

```diff
@@ -203,120 +218,134 @@ MooseVariableFE<OutputType>::add(NumericVector<Number> & residual)
 }
 
 template <typename OutputType>
-const MooseArray<Number> &
+void
+MooseVariableFE<OutputType>::addSolution(const DenseVector<Number> & v) const
+{
+  _element_data->addSolution(_sys.solution(), v);
```

Comment:
Indeed. Let me search a little to see why.

Reply:
Maybe the reference behaves like pointer? Although it is const, it just indicating the reference is not changed but not the content of the reference?

Reply:
Found an answer: *The constness of execute() only affects the this pointer of the class. It makes the type of this a  const T* instead of just T*. This is not a 'deep' const though - it only means the members themselves cannot be changed, but anything they point to or reference still can.* at https://stackoverflow.com/questions/2523516/why-can-i-call-a-non-const-member-function-pointer-from-a-const-method. Not sure it is the standard (although probably).

Reply:
Yes, I'm aware of that, but what is the difference between our use case, and the program below, which emits the compiler error that I expect? (`constness.cpp:25:62: error: binding value of type 'const int' to reference to type 'int' drops 'const' qualifier`)
```c++
#include <memory>

class SystemBase
{
public:
  SystemBase() : _solution(5) {}

  const int & solution() const { return _solution; }

private:
  int _solution;
};

class MooseVariableData
{
public:
  void addSolution(int & solution, int y) const { solution += y; }
};

class MooseVariableFE
{
public:
  MooseVariableFE() { _element_data = std::make_unique<MooseVariableData>(); }

  void addSolution(int y) const { _element_data->addSolution(_sys.solution(), y); }

private:
  std::unique_ptr<MooseVariableData> _element_data;

  SystemBase _sys;
};

int
main()
{
}
```
```

Reply:
Sorry, yea just like the post said; the program below doesn't compile, but if we change to the reference then the program does compile, and the non-const solution getter is called. That's sad as far as ensuring const-correctness goes 😢 
```c++
#include <memory>
#include <iostream>

class SystemBase
{
public:
  SystemBase() : _solution(5) {}
  SystemBase(const SystemBase & copy) : _solution(copy._solution) {}

  const int & solution() const
  {
    std::cout << "Const version\n";
    return _solution;
  }
  int & solution()
  {
    std::cout << "Non-const version\n";
    return _solution;
  }

private:
  int _solution;
};

class MooseVariableData
{
public:
  void addSolution(int & solution, int y) const { solution += y; }
};

class MooseVariableFE
{
public:
  MooseVariableFE(SystemBase & sys) : _sys(sys)
  {
    _element_data = std::make_unique<MooseVariableData>();
  }

  void addSolution(int y) const { _element_data->addSolution(_sys.solution(), y); }

private:
  std::unique_ptr<MooseVariableData> _element_data;

  // SystemBase & _sys;
  SystemBase _sys;
};

int
main()
{
  SystemBase sys;
  MooseVariableFE var(sys);
  var.addSolution(7);
}
```

Reply:
Do you want me to remove the `const` deco for this method?

Reply:
Yes, since it is ultimately modifying a data member (through a reference)